spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpSourceTaskLauncherConfiguration.java

```diff
@@ -0,0 +1,144 @@
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * @author Chris Schaefer
+ */
+@EnableConfigurationProperties({ SftpSourceProperties.class, SftpSourceBatchProperties.class })
+@Import({ SftpSourceRedisIdempotentReceiverConfiguration.class })
+public class SftpSourceTaskLauncherConfiguration {
+	protected static final String SFTP_HOST_PROPERTY_KEY = "sftp_host";
+	protected static final String SFTP_PORT_PROPERTY_KEY = "sftp_port";
+	protected static final String SFTP_USERNAME_PROPERTY_KEY = "sftp_username";
+	protected static final String SFTP_PASSWORD_PROPERTY_KEY = "sftp_password";
+	protected static final String DATASOURCE_URL_PROPERTY_KEY = "spring_datasource_url";
+	protected static final String DATASOURCE_USERNAME_PROPERTY_KEY = "spring_datasource_username";
+
+	private SftpSourceProperties sftpSourceProperties;
+	private SftpSourceBatchProperties sftpSourceBatchProperties;
+
+	@Autowired
+	public SftpSourceTaskLauncherConfiguration(SftpSourceProperties sftpSourceProperties,
+											   SftpSourceBatchProperties sftpSourceBatchProperties) {
+		this.sftpSourceProperties = sftpSourceProperties;
+		this.sftpSourceBatchProperties = sftpSourceBatchProperties;
+	}
+
+	@IdempotentReceiver("idempotentReceiverInterceptor")
+	@Transformer(inputChannel = "sftpFileTaskLaunchChannel", outputChannel = Source.OUTPUT)
+	@ConditionalOnProperty(name = "sftp.taskLauncherOutput")
+	public TaskLaunchRequest sftpFileTaskLauncherTransfomer(Message message) {
+		return new TaskLaunchRequest(sftpSourceBatchProperties.getBatchResourceUri(), getCommandLineArgs(message),
+				getEnvironmentProperties(), getDeploymentProperties(), null);
+	}
+
+	private Map<String, String> getEnvironmentProperties() {
+		Map<String, String> environmentProperties = new HashMap<>();
+		environmentProperties.put(DATASOURCE_URL_PROPERTY_KEY, sftpSourceBatchProperties.getDataSourceUrl());
+		environmentProperties.put(DATASOURCE_USERNAME_PROPERTY_KEY, sftpSourceBatchProperties.getDataSourceUserName());
+		environmentProperties.put(SFTP_HOST_PROPERTY_KEY, sftpSourceProperties.getFactory().getHost());
+		environmentProperties.put(SFTP_USERNAME_PROPERTY_KEY, sftpSourceProperties.getFactory().getUsername());
+		environmentProperties.put(SFTP_PASSWORD_PROPERTY_KEY, sftpSourceProperties.getFactory().getPassword());
+		environmentProperties.put(SFTP_PORT_PROPERTY_KEY, String.valueOf(sftpSourceProperties.getFactory().getPort()));
+
+		String providedProperties = sftpSourceBatchProperties.getEnvironmentProperties();
+
+		if (StringUtils.hasText(providedProperties)) {
+			String[] splitProperties = StringUtils.split(providedProperties, ",");
+			Properties properties = StringUtils.splitArrayElementsIntoProperties(splitProperties, "=");
```

Comment:
Can we do something like `replaceAll("," "/\n")` and then use `new Properties().load(new StringReader(providedProperties))` ?

Reply:
is there a compelling reason to change it? doing so introduces handling a checked exception as well

Reply:
Well, right, no reason indeed. Only the problem that such a logic is done already somewhere and we should just reuse it. 

In some other places it seems for me we accept such a configuration as a `Map<String, String>` and Spring Boot does the trick for us to read properties and return just a map we can manipulate. 
But that approach might requires a breaking change in the format how we request properties from end-user to this matter.

OK. Never mind. Let it live for while! I'm pretty sure you have copied/pasted such a logic from somewhere.

Reply:
i try to keep the copy/paste to a minimum :^)